src/twisted/test/test_twistd.py

```diff
@@ -1911,5 +1956,83 @@ def test_hooksNotCalled(self):
 
 
 
+class ExitWithSignalTests(unittest.TestCase):
+
+    """
+    Tests for L{twisted.application.app._exitWithSignal}.
+    """
+
+    def setUp(self):
+        """
+        Set up the server options and a fake for use by test cases.
+        """
+        self.config = twistd.ServerOptions()
+        self.config.loadedPlugins = {'test_command': MockServiceMaker()}
+        self.config.subOptions = object()
+        self.config.subCommand = 'test_command'
+        self.exitWithSignalCalled = False
+
+        def fakeExitWithSignal(sig):
+            self.exitWithSignalCalled = True
+
+        self.fakeExitWithSignal = fakeExitWithSignal
+
+
+    def test_exitWithSignal(self):
+        """
+        exitWithSignal replaces the existing signal handler with the default
+        handler and sends the replaced signal to the current process.
+        """
+
+        fakeSignalArgs = [None, None]
+
+        def fake_signal(sig, handler):
+            fakeSignalArgs[0] = sig
+            fakeSignalArgs[1] = handler
+
+
+        fakeKillArgs = [None, None]
+
+        def fake_kill(pid, sig):
+            fakeKillArgs[0] = pid
+            fakeKillArgs[1] = sig
+
+        self.patch(signal, 'signal', fake_signal)
+        self.patch(os, 'kill', fake_kill)
+        app._exitWithSignal(signal.SIGINT)
+
+        self.assertEquals(fakeSignalArgs[0], signal.SIGINT)
+        self.assertEquals(fakeSignalArgs[1], signal.SIG_DFL)
+        self.assertEquals(fakeKillArgs[0], os.getpid())
+        self.assertEquals(fakeKillArgs[1], signal.SIGINT)
+
+
+    def test_normalExit(self):
+        """
+        _exitWithSignal is not called if the runner does not exit with a
+        signal.
+        """
+        self.patch(twistd.app, '_exitWithSignal', self.fakeExitWithSignal)
```

Comment:
I didn't make heavy changes in v2 to the test_twistd, so I'm going to have to say, not sure?  The goal here is to make sure _exitWithSignal is called.  Calling the real function would deliver a signal to the test, which would be bad, so it seems like some patching is required with the current implementation.  I could push the patching back to `os.kill()`, which might loosely be considered a platform integration point where mocking is more appropriate, thoughts?

Not saying that's the only way, just the immediately obvious one.  I'll take a look.